Add /status/v1/health client

hakonhall opened the pull request.

hakonhall committed:
Add /status/v1/health client

freva commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/state/HealthResponse.java:
```diff
@@ -0,0 +1,36 @@
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class HealthResponse {
+    @JsonProperty("status")
+    public Status status = new Status();
```
Why is this initialized?

hakonhall replied:
To guarantee non-null status

freva commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/state/State.java:
```diff
@@ -0,0 +1,12 @@
+ */
+public interface State {
+    /** Issue GET on /state/v1/health */
+    HealthResponse getHealth();
```
The return should probably be immutable?

freva commented on node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/configserver/SslConnectionSocketFactoryUpdaterTest.java:
```diff
@@ -0,0 +1,43 @@
+            mock(SslConnectionSocketFactoryCreator.class);
+
+    @Test
+    public void testBasics() {
```
What does this even test?

freva commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/state/HealthResponse.java:
```diff
@@ -0,0 +1,36 @@
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class Status {
+        @JsonProperty("code")
+        public String code;
```
Make this `enum`?

freva commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/configserver/ConfigServerApiImpl.java:
```diff
@@ -48,15 +54,72 @@
+                SSLConnectionSocketFactory.getSocketFactory());
+    }
+
+    public static ConfigServerApiImpl createForTestingWithClient(List<URI> configServerHosts,
```
Not a fan of creating public static methods for testing.. At least one of them can be removed by making the constructor package private, the other can be replaced with `createWithSocketFactory()`

hakonhall replied:
> Not a fan of creating public static methods for testing..

Why?

Used static here to use the same as other ways of constructing it. 

If someone were to try to find a constructor to use by completing "new ConfigServerApiImpl(", they will get no valid (and misleading) suggestions.

> At least one of them can be removed by making the constructor package private, 
> the other can be replaced with createWithSocketFactory()

Fixed

freva replied:
> Why?
> Used static here to use the same as other ways of constructing it.

Yeah, I guess its the `public` I have most issue with, tests can afford to do the setup in a slightly more complicated way rather than have some convenience method that otherwise pollutes the interface.

> If someone were to try to find a constructor to use by completing "new ConfigServerApiImpl(", they will get no valid (and misleading) suggestions.

The remaining method is used by the `ConfigServerApiTest`, so it can be made package private and wont be seen much?

hakonhall replied:
With my version of IntelliJ, it completes to "new ConfigServerApiImpl()" w/error with only private constructors, but to "new ConfigServerApiImpl(" with a parameter list that includes all constructors, even private, if at least one them is non-private. This is trying to complete outside of the package.

hakonhall committed:
Fixes after review round

Also:
 - Handle ConnectException (connection refused) as a synonym for DOWN
 - Add test of StateImpl

hakonhall commented:
PTAL
